File: Leetcode/51.cpp
/**
 * 51. N 皇后
 * link：https://leetcode-cn.com/problems/n-queens/
 */

/**
 *  x & (-x)        ->  lowbit(x)   ->  one of possible position
 *  x & (x - 1)     ->  zero the low bit    -> remove one possible position
 */

class Solution
{
    vector<vector<string>> result;
    vector<int> queens;
    int number;

public:
    vector<vector<string>> solveNQueens(int n)
    {
        queens.resize(n, -1);
        number = n;
        solve(0, 0, 0, 0);
        return result;
    }

    void solve(int row, int columns, int left_diag, int right_diag)
    {
        if (row == number)
        {
            auto board = int2str(queens);
            result.push_back(board);
        }
        else
        {
            int all_pos = ((1 << number) - 1) & (~(columns | left_diag | right_diag));
            while (all_pos != 0)
            {
                int pos = all_pos & (-all_pos);
                all_pos = all_pos & (all_pos - 1);
                int column = __builtin_ctz(pos);
                queens[row] = column;
                solve(row + 1, columns | pos, (left_diag | pos) >> 1, (right_diag | pos) << 1);
                queens[row] = -1;
            }
        }
    }

    vector<string> int2str(vector<int> &queens)
    {
        vector<string> board;
        for (int i = 0; i < number; i++)
        {
            string row = string(number, '.');
            row[queens[i]] = 'Q';
            board.push_back(row);
        }
        return board;
    }
};

File: Leetcode/51.py
# 51. N 皇后
# link：https://leetcode-cn.com/problems/n-queens/

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        path = [-1] * n
        row = ["."] * n

        def generateBoard() -> List[str]:
            board = list()
            for i in range(n):
                row[path[i]] = "Q"
                board.append("".join(row))
                row[path[i]] = "."
            return board

        def solve(row: int, columns: int, left_diag: int, right_diag: int):
            if row == n:
                result.append(generateBoard())
            else:
                all_pos = ((1 << n)-1) & (~(columns | left_diag | right_diag))

                while all_pos != 0:
                    pos = all_pos & (-all_pos)
                    all_pos = all_pos & (all_pos - 1)
                    col = bin(pos-1).count("1")
                    path[row] = col
                    solve(row+1, columns | pos, (left_diag | pos)
                          >> 1, (right_diag | pos) << 1)

        solve(0, 0, 0, 0)
        return result
